src/test/client/nonblocking.cc

```diff
@@ -611,3 +612,589 @@ TEST_F(TestClient, LlreadvLlwritevInvalidFileHandle) {
+    ssize_t bytes_read = 0;
+    rc = client->ll_preadv_pwritev(fh, current_iov++, 1, i * NUM_BUF * 10,
+                                   false, readfinish.get(), &tmpbl);
+    bl.append(tmpbl);
+    ASSERT_EQ(rc, 0);
+    bytes_read = readfinish->wait();
+    ASSERT_EQ(bytes_read, iov_out_non_contiguous[i].iov_len);
+    total_bytes_read += bytes_read; 
+  }
+  ASSERT_EQ(total_bytes_read, bytes_to_write);
+
+  copy_bufferlist_to_iovec(iov_in_non_contiguous, NUM_BUF, &bl, total_bytes_read);
+  for(int i = 0; i < NUM_BUF; ++i) {
+    ASSERT_EQ(0, strncmp((const char*)iov_in_non_contiguous[i].iov_base,
+                         (const char*)iov_out_non_contiguous[i].iov_base,
+                         iov_out_non_contiguous[i].iov_len));
+  }
+
+  client->ll_release(fh);
+  ASSERT_EQ(0, client->ll_unlink(root, filename, myperm));        
+}
+
+TEST_F(TestClient, LlreadvLlwritevBufferUnderflow) {
+  /* Test with large input buffers and small read buffers to make sure I/O
+  happens normally */
+
+  int mypid = getpid();
+  char filename[256];
+
+  client->unmount();
+  TearDown();
+  SetUp();
+
+  sprintf(filename, "test_llreadvllwritevbufferunderflowfile%u", mypid);
+
+  Inode *root, *file;
+  root = client->get_root();
+  ASSERT_NE(root, (Inode *)NULL);
+
+  Fh *fh;
+  struct ceph_statx stx;
+
+  ASSERT_EQ(0, client->ll_createx(root, filename, 0666,
+				  O_RDWR | O_CREAT | O_TRUNC,
+				  &file, &fh, &stx, 0, 0, myperm));
+
+  const int WRITE_MAX = 1024;
+  char out_buf_0[WRITE_MAX], out_buf_1[WRITE_MAX];
+  memset(out_buf_0, 0xcc, sizeof(out_buf_0));
+  memset(out_buf_1, 0xdd, sizeof(out_buf_1));
+  struct iovec iov_out[2] = {
+    {out_buf_0, sizeof(out_buf_0)},
+    {out_buf_1, sizeof(out_buf_1)}
+  };
+
+  ssize_t bytes_to_write = iov_out[0].iov_len + iov_out[1].iov_len;
+
+  // try to read 2x the written data
+  char in_buf_0[WRITE_MAX * 2], in_buf_1[WRITE_MAX * 2];
+  struct iovec iov_in[2] = {
+    {in_buf_0, sizeof(in_buf_0)},
+    {in_buf_1, sizeof(in_buf_1)}
+  };
+
+  std::unique_ptr<C_SaferCond> writefinish = nullptr;
+  std::unique_ptr<C_SaferCond> readfinish = nullptr;
+
+  writefinish.reset(new C_SaferCond("test-nonblocking-writefinish-buffer-underflow"));
+  readfinish.reset(new C_SaferCond("test-nonblocking-readfinish-buffer-underflow"));
+
+  int64_t rc;
+  bufferlist bl;
+
+  rc = client->ll_preadv_pwritev(fh, iov_out, 2, 0, true, writefinish.get(),
+                                 nullptr);
+  ASSERT_EQ(rc, 0);
+  ssize_t bytes_written = writefinish->wait();
+  ASSERT_EQ(bytes_written, bytes_to_write);
+
+  rc = client->ll_preadv_pwritev(fh, iov_in, 2, 0, false, readfinish.get(),
+                                 &bl);
+  ASSERT_EQ(rc, 0);
+  ssize_t bytes_read = readfinish->wait();
+  ASSERT_EQ(bytes_read, bytes_to_write);
+
+  client->ll_release(fh);
+  ASSERT_EQ(0, client->ll_unlink(root, filename, myperm));
+}
+
+TEST_F(TestClient, LlreadvLlwritevWriteOnlyFile) {
```

Comment:
Is this testing if a file with only "w" perms can be written to?

Reply:
yes

Reply:
``` 
ASSERT_EQ(0, client->ll_createx(root, filename, 0666,
				  O_WRONLY | O_CREAT | O_TRUNC,
				  &file, &fh, &stx, 0, 0, myperm));
```